File: length-of-longest-fibonacci-subsequence.cpp
class Solution {
    unordered_map<int,int>mp;
public:
    int solve(int i, int j, vector<int>&arr){
        int target = arr[j] - arr[i];
        if(mp.count(target) && mp[target]<i){
            int k = mp[target];
            return 1 + solve(k,i,arr);
        }
        return 2;
    } 
    int lenLongestFibSubseq(vector<int>& arr) {
        int n = arr.size();
        for(int i=0; i<n; i++){
            mp[arr[i]] = i;
        }
        int res = 0;
        for(int j=0; j<n; j++){
            for(int k=j+1; k<n; k++){
                int length = solve(j,k,arr);
                if(length>=3){
                    res = max(res,length);
                }
            }
        }
        return res;
    }
};

File: length-of-longest-fibonacci-subsequence.py
class Solution:
    def solve(self,j,k,arr,ind)->int:
        target = arr[k]-arr[j]
        if target in ind and ind[target]<j:
            k = ind[target]
            return 1 + self.solve(k,j,arr,ind)
        return 2

    def lenLongestFibSubseq(self, arr: List[int]) -> int:
        ind = {}
        for i in range(len(arr)):
            ind[arr[i]]=i
        res = 0
        for j in range(len(arr)):
            for k in range(j+1,len(arr)):
                length = self.solve(j,k,arr,ind)
                if length>=3:
                    res = max(res,length)
        
        return res
        